### PR TITLE
[Bugfix] Fix V1 GDN state recovery across mixed speculative batches

### PR DESCRIPTION
**Pull request by DimensionSTP**
## Purpose

Fix GDN accepted-state recovery in the V1 model runner when a request transitions from speculative decoding to a regular decode, including batches containing both kinds of requests.

This builds on #40738 and the mixed-batch follow-up by @bhaktatejas922 (`04e09d548739c628b4fe7f66820ac592097074c6`). Their attribution is preserved. This is prepared as a successor, not claimed as an independently invented replacement.

In addition to rebasing those changes, this fixes a remaining partition bug: when every non-speculative request has an accepted-token count of 1, the previous combined implementation discarded the non-speculative counts. The convolution caller then fell back to the counts from the speculative partition. That can provide both the wrong offsets and the wrong number of entries.

The fix keeps the partitions independent even when all non-speculative counts equal 1. It also removes the GPU-value-dependent Python branch used to decide whether to preserve those counts.

The scope is V1 GDN conv/SSM restoration with `mamba_cache_mode=none`. This does not claim to fix every speculative-decoding output difference or to modify the V2 runner.

I utilized AI tools to assist in refining the implementation, executing tests, and analyzing logs.

## Test Plan

- Reproduce the all-one-count partition failure in both request orderings.
- Run the GDN metadata suite and the causal convolution suite.
- Compare CUDA convolution outputs and final history against the PyTorch reference across accepted offsets, sequence lengths, and both cache layouts.
- Call the real GDN forward and compare with manually restored conv/SSM state: ordinary decode, packed decode, and mixed batches; FP32/BF16 SSM states.
- Run Qwen3.5-0.8B plain/ngram decoding at BS1 and BS8.
- Inspect top-5 logprobs at the first residual output divergence.
- Run Ruff and whitespace checks.

## Test Result

- Before the additional fix: 2 new boundary cases failed, 4 passed.
- Metadata and convolution suites: **213 passed**.
- Real GDN forward / state restoration: **12 passed**.
- Ruff check, Ruff format, and `git diff --check`: passed.

Commands from the source checkout in the isolated runtime:

```bash
PYTHONPATH="$PWD" CUDA_VISIBLE_DEVICES=2 ../.venv-main/bin/python -c 'import torch, pytest; torch.backends.cudnn.enabled=False; raise SystemExit(pytest.main(["tests/v1/attention/test_gdn_metadata_builder.py", "tests/kernels/mamba/test_causal_conv1d.py", "-q"]))'
PYTHONPATH="$PWD" CUDA_VISIBLE_DEVICES=3 ../.venv-main/bin/python -m pytest tests/kernels/mamba/test_gdn_spec_state_restore.py -q
```

cuDNN was disabled only for the reference-convolution test process because the local PyTorch reference path segfaulted inside cuDNN. The tested Triton implementation was unchanged by that setting.

Model check: Qwen3.5-0.8B revision `2fc06364715b967f1860aea9cf38778875588b17`, BF16 greedy, seed 2026, TP1, eager, 8 prompts, ngram 5, lookup 2–10, prefill budget 64, 256 max output tokens, prefix caching off.

| Exact plain/ngram token-sequence agreement | Unpatched main | Previous combined patch | This fix |
|---|---:|---:|---:|
| BS1 | 1/8 | 7/8 | 7/8 |
| BS8 | 1/8 | 3/8 | 5/8 |

The two additional speculative-only divergences disappear. The remaining BS8 differences affect the same three requests that also diverge between plain BS1 and BS8. At each first divergent position, the selected tokens are each other's top-2; one execution has an exact tie, while the other has a 0.125 logprob margin. This supports a numerical-ranking explanation but is not a claim of universal token parity.

Environment: main source `f4eccdadefc6501fafeb1a0bf7f171ff24f984b0` plus this change, CUDA SM90, PyTorch 2.13.0+cu129, explicit `VLLM_USE_V2_MODEL_RUNNER=0`, native sampler. Existing official prebuilt extensions were reused; the modified Triton kernels were compiled from this source. Newly rebuilt fused GDN kernels and the V2 runner were not validated by these runs.

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [x] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [x] The test plan, such as providing test command.
- [x] The test results, such as pasting the results comparison before and after, or e2e results
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
</details>

### CODE REVIEW

File: vllm/v1/worker/gpu_model_runner.py [hunk 1]
```diff
@@ -2583,6 +2583,13 @@ def _build_attn_group_metadata(
                     extra_attn_metadata_args["prev_last_scheduled_idx"] = (
                         self.mamba_prev_last_scheduled_idx.gpu[:num_reqs_padded]
                     )
+            elif self.speculative_config is not None and isinstance(
+                builder,
+                (Mamba2AttentionMetadataBuilder, GDNAttentionMetadataBuilder),
+            ):
+                extra_attn_metadata_args = dict(
+                    num_accepted_tokens=self.num_accepted_tokens.gpu[:num_reqs_padded],
+                )
```
[the remark this answers is not in the document]
Comment (DimensionSTP) [reply vllm/v1/worker/gpu_model_runner.py:2595]: Thanks for catching this. Fixed in 95021608d.

The recovery-only branch now uses the same `ubid is None` assertion as the existing speculative branch, rather than passing full-batch accepted counts to a microbatch.

I added eight runner regression cases covering GDN/Mamba2, speculative/recovery-only execution, and execution with/without microbatches. Before the fix, the two recovery-only microbatch cases failed; after the fix, all eight pass, while non-microbatched accepted-count forwarding is preserved.

The combined metadata, runner, causal-convolution, and GDN state-restoration suites pass: **236 passed**. The CUDA suites used the same isolated runtime and process-local cuDNN-reference setting described in the PR.

File: vllm/v1/attention/backends/gdn_attn.py [hunk 2]
```diff
@@ -254,11 +262,34 @@ def build(  # type: ignore[override]
...
+                            num_accepted_tokens,
+                            torch.ones(
+                                num_prefills,
+                                dtype=num_accepted_tokens.dtype,
+                                device=num_accepted_tokens.device,
+                            ),
+                        ]
+                    )
+            else:
+                num_accepted_tokens = None
```
[the remark this answers is not in the document]
Comment (DimensionSTP) [reply vllm/v1/attention/backends/gdn_attn.py:292]: Thanks for flagging this. I traced the current caller and added three regression cases in 9006ca599, but could not reproduce a length mismatch on this path.

`GPUModelRunner` sets `CommonAttentionMetadata.num_reqs` to `num_reqs_padded` and passes `num_accepted_tokens.gpu[:num_reqs_padded]`. For pure regular decode, `split_decodes_and_prefills()` includes the zero-token padded requests in the returned `num_decodes`. Therefore, in this branch, `num_decodes == m.num_reqs == batch_size`; the accepted-count vector is not shortened relative to the state-index vector. The normal input-preparation path also fills accepted counts after the real requests with 1.

The new tests cover (real requests, padded requests) of (1, 4), (3, 8), and (4, 4), using the existing builder unchanged. They verify matching tensor shapes, preserved accepted counts, padding counts of 1, and null source indices for padded rows. All three pass.

I have left the padding implementation unchanged. Could you point me to a caller where this branch receives fewer accepted-count entries than `m.num_reqs`? This check covers the current runner/builder metadata contract, not all CUDA Graph replay behavior.

File: vllm/model_executor/layers/mamba/gdn/qwen_gdn_linear_attn.py [hunk 3]
```diff
@@ -1667,6 +1684,14 @@ def _forward_core_decode_non_spec(
         )
         ssm_state = self_kv_cache[1]
         num_actual_tokens = attn_metadata.num_actual_tokens
+        num_accepted_tokens = attn_metadata.num_accepted_tokens
+        spec_decode_src_indices = attn_metadata.spec_decode_src_indices
+
+        if spec_decode_src_indices is not None:
```
Comment (liulanze) [inline vllm/model_executor/layers/mamba/gdn/qwen_gdn_linear_attn.py:1692]: The ROCm AITER path seems to miss this recovery. `_forward_core_rocm` calls `_forward_core_decode_aiter` for a non-spec decode, but that function does not use `spec_decode_src_indices` or pass `num_accepted_tokens` to the conv kernel. Therefore, after a speculative step, ROCm may still read the wrong SSM and conv states.

Comment (DimensionSTP) [reply vllm/model_executor/layers/mamba/gdn/qwen_gdn_linear_attn.py:1692]: Thanks for catching this. You're right: the AITER decode path could bypass recovery after a speculative step.

I've updated the dispatch condition so batches with spec_decode_src_indices use the existing common path, which restores the SSM state and forwards the accepted-token counts to convolution. Decode batches that do not require recovery retain the AITER fast path.

The dispatch regression fails before the fix. Afterward, all 16 focused CPU checks pass, including state-copy and accepted-token forwarding checks. The CPU tests mock the compute kernels; I haven't validated actual ROCm/AITER execution or its performance.

Could you help validate the speculative-to-non-speculative transition on AMD hardware, or point me to the appropriate AMD CI coverage?

**Comment by DimensionSTP** [comment]
@ZJY0516, could you review the state-recovery approach here and identify any remaining blockers to landing it? This PR has incorporated the ROCm dispatch feedback, and I have now compared it directly with the later alternative in #56531 on main `81d7293c`.

Both fixes handle the mixed-batch and all-zero-draft cases I checked: an accepted count of 3 must select state slot 2 and preserve the convolution offset, while main selects slot 0/count 1. This PR repairs the state when using the non-spec path; #56531 keeps zero-draft decode rows on the speculative path and also changes the V2 runner. My CUDA state-recovery tests pass, including FP32/BF16 conv/SSM reference checks, with 18 focused tests passing overall. The alternative's 4 metadata tests pass too.

The V1 Qwen3.5-0.8B plain/ngram comparison completed all 12 runs across main and both patches. It is not proof of complete equivalence: plain decoding itself changes some outputs across batch sizes. These runs used current Python sources with reused compiled artifacts, and AMD hardware and V2 remain unverified on my side.

I would like to bring this PR to a mergeable decision rather than leave two overlapping fixes pending. Is there a technical objection to landing the recovery approach here, with the additional V2 work handled separately? If retaining the speculative path is the preferred design, I can revise this PR around that direction while preserving the accepted-state regression coverage and coordinating with #56531's author. Please let me know which change or validation would unblock review; I will address it here.

**Review by ZJY0516 — COMMENTED** [review]
We don't want to put too much effort on model runner v1 now

**Comment by DimensionSTP** [comment]
> We don't want to put too much effort on model runner v1 now

@ZJY0516, understood. I am proceeding with a V2-focused rework of #55504 rather than waiting on the current V1-only direction.

By default, I will:

- remove the V1-only production implementation from the final diff;
- adopt the V2 zero-draft routing, preserving the authorship and provenance of #56531;
- retain the distinct FP32/BF16 numerical regression coverage for the actual Qwen GDN convolution and recurrent state; and
- re-run the focused CPU and CUDA validation on current main.

If your intent is to preserve the already implemented and validated V1 recovery and review a unified V1+V2 change instead, I can retain the existing V1 coverage without adding further V1-specific scope and integrate the V2 path into the same PR.

Unless you indicate that the unified V1+V2 scope is preferred, I will proceed with the V2-focused update in #55504 and explicitly document and reconcile its overlap with #56531 so the implementation and state-level regression coverage can be reviewed together.

**Comment by DimensionSTP** [comment]
@ZJY0516, the V2-focused rework I mentioned is now pushed to this PR. The final diff contains only the V2 routing change and three regression-test files; the V1-only implementation is no longer in the final diff.

The runtime change is adapted from #56531 by @vadiklyutiy. This PR also adds FP32/BF16 CUDA checks of the actual GDN convolution and recurrent state at the accepted-token position. On the current main used for the update, the focused worker and zero-draft/CUDA tests passed. Two failures in the broader test run also reproduce on unmodified main.

Could you review the V2-focused diff and let me know if anything blocks landing #55504? The upstream pre-commit job was skipped by the external-contributor pre-run gate, so a maintainer-triggered CI run would also help.
